app/controllers/display_ads_controller.rb

```diff
@@ -0,0 +1,30 @@
+class DisplayAdsController < ApplicationController
+  before_action :set_cache_control_headers, only: %i[for_display], unless: -> { current_user }
+  CACHE_EXPIRY_FOR_DISPLAY_ADS = 15.minutes.to_i.freeze
+
+  def for_display
+    skip_authorization
+    set_cache_control_headers(CACHE_EXPIRY_FOR_DISPLAY_ADS) unless session_current_user_id
+
+    if params[:placement_area]
+      if params[:article_id].present?
+        @article = Article.find(params[:article_id])
+      end
+
+      @display_ad = DisplayAd.for_display(
+        area: params[:placement_area],
+        user_signed_in: user_signed_in?,
+        organization_id: @article&.organization_id,
+        permit_adjacent_sponsors: ArticleDecorator.new(@article).permit_adjacent_sponsors?,
+        article_tags: @article&.decorate&.cached_tag_list_array || [],
+        article_id: @article&.id,
+      )
+
+      if @display_ad && !session_current_user_id
+        set_surrogate_key_header @display_ad.record_key
+      end
+    end
+
+    render layout: false
```

Comment:
So, another take-or-leave idea here, but, given that we're rendering without a layout, and given that the template's only "job" is to conditionally wrap a partial in cache, and given that we might rather fewer conditional code in views, could we maybe move the cache conditional into the controller?

Something like:

```suggestion
    render plain: "" and return unless @display_ad

    caching do
      render partial: "shared/display_ad",
             locals: { display_ad: @display_ad,
			                 data_context_type: DisplayAdEvent::CONTEXT_TYPE_ARTICLE }
    end
  end

  private

  def caching(&block)
    if user_signed_in?
      yield
    else
      cache(["display_ad", @article, @display_ad.placement_area], expires_in: 15.minutes, &block)
    end
```

Reply:
Somewhat separately: this setup with `cache [@article, @display_ad.placement_area]` somewhat reinforces my thinking elsewhere about the routing hierarchy, but more importantly, seems unlikely to work in the potential-future we're working elsewhere to leave room for, where `/display_ads` is an end-point available for other non-article placement-areas as well. I think we _might_ make this more flexible if it used `params` instead — `cache "display_ad", params[:slug], params[:placement_area]` possibly, pending other things — as those should collapse if empty, without raising exceptions.

Reply:
@jaw6 thank you so much for this! I learned a bit from this snippet as well. However, I decided to keep the view instead of moving it in the controller because I find it a bit more readable. I'm not sure if that was the right decision but maybe at some point we can re-evaluate it if more folks feel that having it in the controller is more intuitive.

> I think we might make this more flexible if it used params instead — cache "display_ad", params[:slug], params[:placement_area] possibly, pending other things — as those should collapse if empty, without raising exceptions.

Thanks this makes sense, I've updated accordingly, this is what it looks like now. I also added the username and the slug for consistency.

```
Read fragment views/display_ads/show:17d892e28bb88b0c9a9c2c445c89ef79/org6121/post-on-behalf-of-herber-org-4ajc/post_comments (2.0ms)
```